File: CSharpCodeBase/utils/ailib/aisequence.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'

local AISequenceNode = Class({__includes=AINode, init = function (self, entity, childNodes)
      AINode.init(self, entity, childNodes)
      self.idx = 1
end})

function AISequenceNode:Reset()
  AINode.Reset(self)
  self.idx = 1
end

function AISequenceNode:Visit()
  if self.status ~= NODE_RUNNING then
    self.idx = 1
  end
  
  while self.idx <= #self.childNodes do
    local childNode = self.childNodes[self.idx]
    local childStatus = NODE_FAILURE
    if not childNode:IsSleeping() then
      childNode:Visit()
      childStatus = childNode.status
    end
    if childStatus == NODE_RUNNING or childStatus == NODE_FAILURE then
      if childNode:IsSleeping() then
        self.status = NODE_FAILURE
      else
        self.status = childNode.status
      end
      return
    end
    self.idx = self.idx + 1
  end
  
  self.status = NODE_SUCCESS
end

return AISequenceNode

File: CSharpCodeBase/utils/ailib/aisequence.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class aisequence :ainode{
 var AISequenceNode = Class({__includes=AINode, init = public void (self, entity, childNodes){
       AINode.init(self, entity, childNodes);
       this.idx = 1;
 }})
 public void AISequenceNode:Reset(){
   AINode.Reset(self);
   this.idx = 1;
 }
 public void AISequenceNode:Visit(){
   if(this.status != NODE_RUNNING  ){ 
     this.idx = 1;
   }
   
   while(this.idx <= #this.childNodes ){
     var childNode = this.childNodes[this.idx];
     var childStatus = NODE_FAILURE;
     if(not childNode:IsSleeping()  ){ 
       childNode:Visit();
       childStatus = childNode.status;
     }
     if(childStatus == NODE_RUNNING  ||  childStatus == NODE_FAILURE  ){ 
       if(childNode:IsSleeping()  ){ 
         this.status = NODE_FAILURE;
       }else{
         this.status = childNode.status;
       }
       return;
     }
     this.idx = this.idx + 1;
   }
   
   this.status = NODE_SUCCESS;
 }
 Node}}